common/workunit/workunit.cpp

```diff
@@ -5356,28 +5356,47 @@ void CWorkUnitFactory::importWorkUnit(const char *zapReportFileName, const char
 {
     class CImportWorkUnitHelper
     {
-        StringAttr zapReportFileName, zapReportPassword, importDir, user, wuid, fromWUID, scope;
+        StringAttr zapReportFileName, zapReportPassword, importDir, user, wuid, fromWUID, scope, zapFolder;
         StringBuffer unzipDir;
         Owned<IPTree> wuTree, graphProgressTree;
 
-        bool findZAPFile(const char *mask, bool optional, StringBuffer &fileName)
+        //To keep the file layout of the post-mortem info, the '-r' option is used to create a ZAP
+        //report. When the ZAP report is unzipped, an extra folder layer is added on the top of the
+        //ZAP folder. The findZAPFolder() is added to find out the ZAP folder.
+        void findZAPFolder()
         {
             Owned<IFile> d = createIFile(unzipDir);
             if (!d->exists())
-                throw MakeStringException(WUERR_InvalidUserInput, "%s not found.", unzipDir.str());
+                throw makeStringExceptionV(WUERR_CannotImportWorkunit, "%s not exist.", unzipDir.str());
+
+            Owned<IDirectoryIterator> zapFiles = d->directoryFiles("ZAPReport_W*.xml", false, false);
+            if (zapFiles->first()) //The ZAP folder should has a ZAPReport_W*.xml.
+            { //legacy ZAP report: the ZAPReport_W*.xml is in unzipDir.
+                zapFolder.set(unzipDir);
+                return;
+            }
 
+            //The name of the ZAP folder starts with 'ZAPReport_'.
+            Owned<IDirectoryIterator> zapFolders = d->directoryFiles("ZAPReport_*", false, true);
+            if (!zapFolders->first())
+                throw makeStringExceptionV(WUERR_CannotImportWorkunit, "Failed to find ZAP folder in %s.", unzipDir.str());
+            zapFolder.set(zapFolders->query().queryFilename());
+        }
+        bool findZAPFile(const char *mask, bool optional, StringBuffer &fileName)
```

Comment:
The naming of findZAPFolder vs findZAPFile is a bit confusing.
Former is meant to find a sub-folder with the extracted zap in the working folder.
findZAPFile which is meant to find a file within the actual [extracted] zap.

Depending on whether findZAPFolder remains., findZAPFile, might be better named findFileInZAPFolder..

future: I wonder if it wouldn't be better to have a 'type' param that explicitly built the correct pattern for the file(s) it was looking for, since some of them are no longer going to be at the same top directory level.

Reply:
Will rename